### ISSUE TITLE
connection timeouts in body_sync (requesting blocks)

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#454]
Seeing a node stuck trying to ask for more blocks and repeatedly timing out.

On the surface this looks suspiciously like this was caused by us Increasing the # blocks being requested each body_sync iteration.

```
Dec 10 02:39:22.453 DEBG block_sync: 19597/19785 requesting blocks [1bbd740d, 31703d3d, 243f79e6, ba833426, 64142ed6, 5bf2e0a6, 016bf1d0, a07a4e67, ae19c4e5, afbc415c, 949950b7, 4333497d, 345d97d6, 0146e928, e6d24ebc, 3975450d, 77c6f308, d423be34, b018b00c, f60fa457] from 2 peers
Dec 10 02:39:22.454 DEBG Requesting block 1bbd740d from peer 37.6.168.203:23414.
Dec 10 02:39:22.454 DEBG Requesting block 31703d3d from peer 37.6.168.203:23414.
Dec 10 02:39:22.455 DEBG Requesting block 243f79e6 from peer 37.6.168.203:23414.
Dec 10 02:39:22.455 DEBG Requesting block ba833426 from peer 37.6.168.203:23414.
Dec 10 02:39:22.456 DEBG Requesting block 64142ed6 from peer 37.6.168.203:23414.
Dec 10 02:39:22.456 DEBG Requesting block 5bf2e0a6 from peer 37.6.168.203:23414.
Dec 10 02:39:22.457 DEBG Requesting block 016bf1d0 from peer 51.15.219.12:13414.
Dec 10 02:39:22.457 DEBG Requesting block a07a4e67 from peer 51.15.219.12:13414.
Dec 10 02:39:22.458 DEBG Requesting block ae19c4e5 from peer 37.6.168.203:23414.
Dec 10 02:39:22.458 DEBG Requesting block afbc415c from peer 37.6.168.203:23414.
Dec 10 02:39:22.458 DEBG Requesting block 949950b7 from peer 51.15.219.12:13414.
Dec 10 02:39:22.459 DEBG Requesting block 4333497d from peer 37.6.168.203:23414.
Dec 10 02:39:22.459 DEBG Requesting block 345d97d6 from peer 51.15.219.12:13414.
Dec 10 02:39:22.460 DEBG Requesting block 0146e928 from peer 37.6.168.203:23414.
Dec 10 02:39:22.460 DEBG Requesting block e6d24ebc from peer 37.6.168.203:23414.
Dec 10 02:39:22.461 DEBG Requesting block 3975450d from peer 37.6.168.203:23414.
Dec 10 02:39:22.461 DEBG Requesting block 77c6f308 from peer 51.15.219.12:13414.
Dec 10 02:39:22.462 DEBG Requesting block d423be34 from peer 51.15.219.12:13414.
Dec 10 02:39:22.462 DEBG Requesting block b018b00c from peer 51.15.219.12:13414.
Dec 10 02:39:22.462 DEBG Requesting block f60fa457 from peer 37.6.168.203:23414.
Dec 10 02:39:24.348 TRCE Too long: Block Some(1bbd740d)
Dec 10 02:39:24.348 INFO Client 37.6.168.203:23414 connection lost: Timeout
```

Note: timeout occurs on the 1st block being requested.
Note: asking for 20 blocks from 2 peers.

I'm wondering what happens on the receiving end of this when a single peer requests multiple blocks from another peer.
Are these handled sequentially by the peer connection on the receiving end?
I'm having trouble understanding what happens at the connection level when we're using the same peer-peer connection to make multiple "parallel" requests like this.

I would have expected the last few to timeout - that the first few were handled within 5s but some of the later ones did not.
But this is not the case here where we see the timeout on the 1st request.

Alternatively what we're seeing here is the backlog from previous body_sync loop?
Maybe we're still processing block requests from a previous sync loop causing these new ones to timeout without ever having a chance to run.

**Comment by antiochp** [comment]
See #477 for proposed solution to the "interleaving" issue.

**Comment by ignopeverell** [comment]
Fixed by #479
